compiler/rustc_middle/src/ty/consts/int.rs

```diff
@@ -109,3 +112,258 @@ impl std::fmt::Debug for ConstInt {
+            truncate(self.data, self.size()),
+            { self.data },
+            "Scalar value {:#x} exceeds size of {} bytes",
+            { self.data },
+            self.size
+        );
+    }
+
+    #[inline]
+    pub fn zst() -> Self {
+        Self::null(Size::ZERO)
+    }
+
+    #[inline]
+    pub fn null(size: Size) -> Self {
+        Self { data: 0, size: size.bytes() as u8 }
+    }
+
+    pub(crate) fn ptr_sized_op<'tcx>(
+        self,
+        dl: &TargetDataLayout,
+        f_int: impl FnOnce(u64) -> InterpResult<'tcx, u64>,
+    ) -> InterpResult<'tcx, Self> {
+        assert_eq!(u64::from(self.size), dl.pointer_size.bytes());
+        Ok(Self { data: u128::from(f_int(u64::try_from(self.data).unwrap())?), size: self.size })
+    }
+
+    #[inline]
+    pub fn try_from_uint(i: impl Into<u128>, size: Size) -> Option<Self> {
+        let data = i.into();
+        if truncate(data, size) == data {
+            Some(Self { data, size: size.bytes() as u8 })
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn try_from_int(i: impl Into<i128>, size: Size) -> Option<Self> {
+        let i = i.into();
+        // `into` performed sign extension, we have to truncate
+        let truncated = truncate(i as u128, size);
+        if sign_extend(truncated, size) as i128 == i {
+            Some(Self { data: truncated, size: size.bytes() as u8 })
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn assert_bits(self, target_size: Size) -> u128 {
+        assert_ne!(target_size.bytes(), 0, "you should never look at the bits of a ZST");
+        assert_eq!(target_size.bytes(), u64::from(self.size));
+        self.check_data();
+        self.data
+    }
+
+    #[inline]
+    pub fn to_bits(self, target_size: Size) -> InterpResult<'static, u128> {
+        assert_ne!(target_size.bytes(), 0, "you should never look at the bits of a ZST");
+        if target_size.bytes() != u64::from(self.size) {
+            throw_ub!(ScalarSizeMismatch {
+                target_size: target_size.bytes(),
+                data_size: u64::from(self.size),
+            });
+        }
+        self.check_data();
+        Ok(self.data)
+    }
+}
+
+macro_rules! from {
+    ($($ty:ty),*) => {
+        $(
+            impl From<$ty> for ScalarInt {
+                #[inline]
+                fn from(u: $ty) -> Self {
+                    Self {
+                        data: u128::from(u),
+                        size: std::mem::size_of::<$ty>() as u8,
+                    }
+                }
+            }
+        )*
+    }
+}
+
+macro_rules! try_from {
+    ($($ty:ty),*) => {
+        $(
+            impl TryFrom<ScalarInt> for $ty {
+                type Error = InterpErrorInfo<'static>;
+                #[inline]
+                fn try_from(int: ScalarInt) -> InterpResult<'static, Self> {
+                    int.to_bits(Size::from_bytes(std::mem::size_of::<$ty>())).map(|u| u.try_into().unwrap())
```

Comment:
This is a rather strange function... on failure it will allocate a full CTFE error (including backtrace!) and then throw it away again. I don't think we should do that.

What is this used for anyway? I think it is new.

Reply:
we are not allocating a CTFE error and throwing it away. If `to_bits` fails, we forward that error outwards. We are throwing away the `<u128 as TryInto<$ty>>::Error` with the `unwrap()`, which is not really throwing it away, but putting it into a panic message.

These impls here are only used for the impls below, which are used in pretty printing.

Reply:
Oh right, `InterpErrorInfo` is the packaged error.

But if this is for pretty-printing, the caller will throw the error away. I don't think this should be throwing interpreter errors -- this is less part of the interpreter and more part of the interface between the interpreter and the rest of the compiler.